modules/packages/LinearAlgebra.chpl

```diff
@@ -221,6 +222,260 @@ class LinearAlgebraError : Error {
     }
 }
 
+/*
+This class promotes lazy computation of required
+Helper matrices to evaluate matrix exponential.
+
+High precision of expm requires smaller computations,
+hence we do a lazy computation depending on the norm value
+of the matrix.
+
+*/
+class ExpmPadeHelper {
+  type t;
+  var A : t;
+  var A2 : t;
+  var A4 : t;
+  var A6 : t;
+  var A8 : t;
+  var A10 : t;
+  var Exactd4 : real;
+  var Exactd6 : real;
+  var Exactd8 : real;
+  var Exactd10 : real;
+  var Approxd4 : real;
+  var Approxd6 : real;
+  var Approxd8 : real;
+  var Approxd10 : real;
+  var ident : t;
+  var use_exact_onenorm : bool;
+  // Boolean arrays to know if A, Exactd, Approxd
+  // have been computed.
+  var isAComputed : [1..5] bool;
+  var isExactdComputed : [1..4] bool;
+  var isApproxdComputed : [1..4] bool;
+
+  /*
+  Params:
+  A: Expects an N*N square matrix
+  use_exact_onenorm: boolean value specifying if the onenorm has to be exact.
+
+  TODO: Need to write a method to find approximate norm.
+  Currently only supports for use_exact_onenorm = true.
+  */
+  proc init(A, use_exact_onenorm: bool) {
+    this.t = A.type;
+    this.A = A;
+    this.ident = eye(A.indices);
+    this.use_exact_onenorm = use_exact_onenorm;
+  }
+
+  proc comp_A2(){
+    if !isAComputed[1] then {
+      this.A2 = matrix_prod(this.A, this.A);
+      isAComputed[1] = true;
+    }
+    return this.A2;
+  }
+
+  proc comp_A4(){
+    if !isAComputed[2] then {
+      this.A4 = matrix_prod(comp_A2(), comp_A2());
+      isAComputed[2] = true;
+    }
+    return this.A4;
+  }
+
+  proc comp_A6(){
+    if !isAComputed[3] then {
+      this.A6 = matrix_prod(comp_A4(), comp_A2());
+      isAComputed[3] = true;
+    }
+    return this.A6;
+  }
+
+  proc comp_A8(){
+    if !isAComputed[4] then {
+      this.A8 = matrix_prod(comp_A6(), comp_A2());
+      isAComputed[4] = true;
+    }
+    return this.A8;
+  }
+
+  proc comp_A10(){
+    if !isAComputed[5] then {
+      this.A10 = matrix_prod(comp_A8(), comp_A2());
+      isAComputed[5] = true;
+    }
+    return this.A10;
+  }
+
+  proc matrix_prod(A: [], B: []){
```

Comment:
Is there a reason to abstract this into its own function? It would seem simpler to just call `dot`.

Reply:
Hi @npadmana, I was intending to implement and use strassen approach to further optimize it.

Reply:
My suggestion is that we don't get into worrying about matrix multiplication. That's a can of worms that will completely take over the summer! :-)